### PR TITLE
Gradient matching (Witches' Brew) for Tensorflow

### LINKED ISSUE #1386: Implement Witches’ brew  poisoning attack
**Is your feature request related to a problem? Please describe.**
Implement poisoning attack: 
`Witches' Brew: Industrial Scale Data Poisoning via Gradient Matching` 
https://arxiv.org/abs/2009.02276 

**Describe the solution you'd like**

- Jupyter notebook illustrating the use of the attack 
- Example folder 
- Implementation for Pytorch and Keras framework.

**Describe alternatives you've considered**
NA

**Additional context**
NA

### PR DESCRIPTION
**Pull request by TS-Lee**
# Description

This branch adds Gradient Matching (Witches' Brew) poisoning attack. This attack computes the clean label noise by setting the training gradient of poisoned samples towards the poisoning target class. This PR is for a TF version. The current commit is work in progress, and the follow up will come with a test and an example notebook.

Fixes #1386 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS: Ubuntu 18.02
- Python version: 3.8
- ART version or commit number: dev_1.9.0
- TensorFlow / Keras / PyTorch / MXNet version: Tensorflow 2.4.1.

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+                super().__init__(max_value=max_value)
+
+            def __call__(self, w):
+                return tf.clip_by_value(w, -self.max_value, self.max_value)
+
+        # Define the model to apply and optimize the poison.
+        input_poison = Input(batch_shape=self.substitute_classifier.model.input.shape)
+        input_indices = Input(shape=())
+        # y_true_poison = Input(shape=self.substitute_classifier.model.output.shape)
+        # y_true_poison = Input(shape=np.shape(y_trigger)[1:])
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:207]: Are these commented lines still needed?

File: tests/attacks/poison/test_gradient_matching_attack.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Comment (beat-buesser) [inline tests/attacks/poison/test_gradient_matching_attack.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2022
```

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2021
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2022
```

File: notebooks/poisoning_attack_witches_brew.ipynb [hunk 4]
```diff
@@ -0,0 +1,1098 @@
+{
```
Comment (beat-buesser) [inline notebooks/poisoning_attack_witches_brew.ipynb]: Could you please update the notebooks/README with a link and abstract for this new notebook?

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+                return {"learning_rates": self.learning_rates, "milestones": self.milestones}
+
+        lr_schedule = tf.keras.callbacks.LearningRateScheduler(PredefinedLRSchedule(*self.learning_rate_schedule))
+
+        class SignedAdam(tf.keras.optimizers.Adam):
+            """
+            This optimizer takes only the sign of the gradients and pass it to the Adam optimizer.
+            """
+
+            def compute_gradients(
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:259]: Please add typing.

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+
+        class PredefinedLRSchedule(tf.keras.optimizers.schedules.LearningRateSchedule):
+            """
+            Use a preset learning rate based on the current training epoch.
+            """
+
+            def __init__(self, learning_rates, milestones):
+                self.schedule = list(zip(milestones, learning_rates))
+
+            def __call__(self, step):
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:241]: Please add typing.

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+
+        model_trainable = self.substitute_classifier.model.trainable
+        self.substitute_classifier.model.trainable = False
+
+        class PredefinedLRSchedule(tf.keras.optimizers.schedules.LearningRateSchedule):
+            """
+            Use a preset learning rate based on the current training epoch.
+            """
+
+            def __init__(self, learning_rates, milestones):
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:238]: Please add typing.

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+        embedding_layer = Embedding(len(x_poison), np.prod(input_poison.shape[1:]))
+        embeddings = embedding_layer(input_indices)
+        embeddings = ClipConstraint(max_value=self.epsilon)(embeddings)
+        embeddings = tf.reshape(embeddings, tf.shape(input_poison))
+        input_noised = Add()([input_poison, embeddings])
+        input_noised = Lambda(lambda x: K.clip(x, self.clip_values[0], self.clip_values[1]))(
+            input_noised
+        )  # Make sure the poisoned samples are in a valid range.
+
+        def loss_fn(input_noised, target, grad_ws_norm):
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:218]: Please add typing.

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+            return d_w_norm
+
+        grad_ws_norm = grad_loss(self.substitute_classifier.model, tf.constant(x_trigger), tf.constant(y_trigger))
+
+        class ClipConstraint(tf.keras.constraints.MaxNorm):
+            """
+            Clip the tensor values.
+            """
+
+            def __init__(self, max_value=2):
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:197]: Please add typing.

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+                super().__init__(max_value=max_value)
+
+            def __call__(self, w):
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:200]: Please add typing.

File: art/attacks/poisoning/gradient_matching_attack.py (hunk 1 above)
Comment (TS-Lee) [reply art/attacks/poisoning/gradient_matching_attack.py:207]: Not needed. I already removed those in my branch.. I will update the PR soon.

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 11]
```diff
@@ -0,0 +1,556 @@
...
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements Gradient Matching clean-label attacks (a.k.a. Witches' Brew) on Neural Networks.
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:19]:
```suggestion
This module implements Gradient Matching clean-label attacks (a.k.a. Witches' Brew) on Neural Networks.

| Paper link: https://arxiv.org/abs/2009.02276
```

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 12]
```diff
@@ -0,0 +1,556 @@
...
+
+        # def grad(classifier: "CLASSIFIER_NEURALNETWORK_TYPE", x: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        #     classifier.model.zero_grad()
+        #     y = classifier.model(x)
+        #     loss_ = classifier.loss(y, target)
+        #     loss_.backward()
+        #     d_w = [w.grad for w in classifier.model.parameters()]
+        #     d_w = torch.cat([w.flatten() for w in d_w])
+        #     d_w_norm = d_w / torch.sqrt(torch.sum(torch.square(d_w)))
+        #     return d_w_norm
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:248]: Is this commented code still required?

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 13]
```diff
@@ -0,0 +1,556 @@
...
+            if B_ < best_B:
+                best_B = B_  # pylint: disable=C0103
+                best_x_poisoned = x_poisoned
+                best_indices_poison = indices_poison
+
+        x_train[best_indices_poison] = best_x_poisoned
+        return x_train, y_train  # y_train has not been modified.
+
+    def __poison__pytorch(
+        self, x_trigger: np.ndarray, y_trigger: np.ndarray, x_poison: np.ndarray, y_poison: np.ndarray  # pylint: disable=unused-argument
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:446]: Why is there an unused argument?

File: art/attacks/poisoning/gradient_matching_attack.py [hunk 14]
```diff
@@ -0,0 +1,556 @@
...
+        B, poisoned_samples = self.backdoor_model(  # pylint: disable=C0103
+            torch.tensor(x_poison, device=device, dtype=torch.float),
+            torch.arange(0, len(x_poison), dtype=torch.int32),
+            torch.tensor(y_poison, device=device, dtype=torch.long),
+            self.grad_ws_norm,
+        )
+        return B.detach().numpy(), poisoned_samples.detach().numpy()
+
+    def __poison__tensorflow(
+        self, x_trigger: np.ndarray, y_trigger: np.ndarray, x_poison: np.ndarray, y_poison: np.ndarray  # pylint: disable=unused-argument
```
Comment (beat-buesser) [inline art/attacks/poisoning/gradient_matching_attack.py:501]: Why is there an unused argument?

File: notebooks/README.md [hunk 15]
```diff
@@ -216,6 +216,10 @@ demonstrates a defense against poisoning attacks that generation the suspected b
 [poisoning_defence_strip.ipynb](poisoning_defence_strip.ipynb) [[on nbviewer](https://nbviewer.jupyter.org/github/Trusted-AI/adversarial-robustness-toolbox/blob/main/notebooks/poisoning_defence_strip.ipynb)]
 demonstrates a defense against input-agnostic backdoor attacks that filters suspicious inputs at runtime.
 
+[poisoning_attack_witches_brew.ipynb](poisoning_attack_witches_brew.ipynb) [[on nbviewer](https://nbviewer.jupyter.org/github/Trusted-AI/adversarial-robustness-toolbox/blob/main/notebooks/poisoning_attack_witches_brew.ipynb)]
+demonstrates the gradient matching poisoning attack (a.k.a. Witches' Brew) that adds noise to align the training gradient to a specific direction that can poison the target model.
```
Comment (beat-buesser) [inline notebooks/README.md:220]:
```suggestion
Demonstrates the gradient matching poisoning attack (a.k.a. Witches' Brew) that adds noise to align the training gradient to a specific direction that can poison the target model.
```

File: tests/attacks/poison/test_gradient_matching_attack.py [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        x_trigger = x_test[index_target : index_target + 1]
+
+        x_train, y_train = x_train[:1000], y_train[:1000]
+        y_train = np.argmax(y_train, axis=-1)
+        attack = GradientMatchingAttack(
+            classifier, epsilon=0.3, percent_poison=0.01, max_trials=1, max_epochs=1, verbose=False
+        )
+        x_poison, y_poison = attack.poison(x_trigger, [class_target], x_train, y_train)
+
+        np.testing.assert_equal(np.shape(x_poison), np.shape(x_train))
```
Comment (beat-buesser) [inline tests/attacks/poison/test_gradient_matching_attack.py:49]: Would it be possible to add additional checks to see if the poisoned data is different from the benign data?

File: tests/attacks/poison/test_gradient_matching_attack.py [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+import numpy as np
+import pytest
+
+from art.attacks.poisoning import GradientMatchingAttack
+
+from tests.utils import ARTTestException
+
+logger = logging.getLogger(__name__)
+
+
```
Comment (beat-buesser) [inline tests/attacks/poison/test_gradient_matching_attack.py:30]: Could you please add a test that covers the exceptions in `_check_params` following the pattern in similar test modules?

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @TS-Lee Thank you very much for your pull request implementing the Gradient Matching poisoning attack! I have added a few more review comments, let me know what you think.

**Comment by beat-buesser** [comment]
About the notebook:
- Could you please suppress the TensorFlow warnings
- Could you pleas check the last cell? It seems that it does not print the final results.

**Comment by TS-Lee** [comment]
Let me use another branch for easier integration please. It branched out two commits before this.
I created a new PR with the branch: https://github.com/Trusted-AI/adversarial-robustness-toolbox/pull/1575

File: notebooks/README.md (hunk 15 above)
Comment (TS-Lee) [reply notebooks/README.md:220]: I'm not sure about this change. I see all other lines use this capitalization. Or did I miss some other change?

File: art/attacks/poisoning/gradient_matching_attack.py (hunk 14 above)
Comment (TS-Lee) [reply art/attacks/poisoning/gradient_matching_attack.py:596]: Fixed. It is a remnant from the previous method definition when it was overriding poison.

File: tests/attacks/poison/test_gradient_matching_attack.py (hunk 16 above)
Comment (TS-Lee) [reply tests/attacks/poison/test_gradient_matching_attack.py:49]: Added a couple to check epsilon and percent_poison.

File: art/attacks/poisoning/gradient_matching_attack.py (hunk 13 above)
Comment (TS-Lee) [reply art/attacks/poisoning/gradient_matching_attack.py:514]: Fixed.

File: art/attacks/poisoning/gradient_matching_attack.py (hunk 12 above)
Comment (TS-Lee) [reply art/attacks/poisoning/gradient_matching_attack.py:282]: removed.

File: art/attacks/poisoning/gradient_matching_attack.py (hunk 11 above)
Comment (TS-Lee) [reply art/attacks/poisoning/gradient_matching_attack.py:19]: Added the link.
